File: Plugins/slua_unreal/Source/slua_profile/Private/slua_profile.cpp
// Tencent is pleased to support the open source community by making sluaunreal available.

// Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
// Licensed under the BSD 3-Clause License (the "License"); 
// you may not use this file except in compliance with the License. You may obtain a copy of the License at

// https://opensource.org/licenses/BSD-3-Clause

// Unless required by applicable law or agreed to in writing, 
// software distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and limitations under the License.

#include "slua_profile.h"
#include "Framework/MultiBox/MultiBoxBuilder.h"
#include "HAL/FileManager.h"
#include "Containers/Ticker.h"
#include "LuaState.h"
#if WITH_EDITOR
#include "LevelEditor.h"
#include "EditorStyle.h"
#include "EditorStyleSet.h"
#endif

#include "LuaProfiler.h"
#include "slua_remote_profile.h"
#include "ProfileDataDefine.h"
#include "SluaProfilerDataManager.h"

DEFINE_LOG_CATEGORY(LogSluaProfile)
#define LOCTEXT_NAMESPACE "Fslua_profileModule"
#define ROOT_NAME "ROOT"
namespace {
    static const FName slua_profileTabName(TEXT("slua_profile"));

    typedef TMap<int64, NS_SLUA::LuaMemInfo> LuaMemInfoMap;
    LuaMemInfoMap memoryInfo;

    typedef TQueue<MemoryFramePtr, EQueueMode::Mpsc> MemoryFrameQueue;
    MemoryFrameQueue memoryQueue;
    MemoryFramePtr currentMemory = MakeShareable(new MemoryFrame());

    ProfileCallInfoArray profilerStack;
    ProfileNodePtr funcProfilerRoot;
    TQueue<ProfileNodePtr, EQueueMode::Mpsc> funcProfilerNodeQueue;
}

void Fslua_profileModule::StartupModule()
{
#if WITH_EDITOR
    Flua_profileCommands::Register();

    PluginCommands = MakeShareable(new FUICommandList);

    PluginCommands->MapAction(
        Flua_profileCommands::Get().OpenPluginWindow,
        FExecuteAction::CreateRaw(this, &Fslua_profileModule::PluginButtonClicked),
        FCanExecuteAction());

    FLevelEditorModule& LevelEditorModule = FModuleManager::LoadModuleChecked<FLevelEditorModule>("LevelEditor");

    {
        TSharedPtr<FExtender> MenuExtender = MakeShareable(new FExtender());
        MenuExtender->AddMenuExtension("WindowLayout", EExtensionHook::After, PluginCommands, FMenuExtensionDelegate::CreateRaw(this, &Fslua_profileModule::AddMenuExtension));

        LevelEditorModule.GetMenuExtensibilityManager()->AddExtender(MenuExtender);
    }


    if (GIsEditor && !IsRunningCommandlet())
    {
        sluaProfilerInspector = MakeShareable(new SProfilerInspector);
        FGlobalTabmanager::Get()->RegisterNomadTabSpawner(slua_profileTabName,
            FOnSpawnTab::CreateRaw(this, &Fslua_profileModule::OnSpawnPluginTab))
            .SetDisplayName(LOCTEXT("Flua_wrapperTabTitle", "slua Profiler"))
            .SetMenuType(ETabSpawnerMenuType::Hidden);
        TickDelegate = FTickerDelegate::CreateRaw(this, &Fslua_profileModule::Tick);
#if ENGINE_MAJOR_VERSION==5
        TickDelegateHandle = FTSTicker::GetCoreTicker().AddTicker(TickDelegate);
#else
        TickDelegateHandle = FTicker::GetCoreTicker().AddTicker(TickDelegate);
#endif
    }

#endif
}

void Fslua_profileModule::ShutdownModule()
{
#if WITH_EDITOR
    sluaProfilerInspector = nullptr;
    ClearCurProfiler();

    Flua_profileCommands::Unregister();

    FGlobalTabmanager::Get()->UnregisterNomadTabSpawner(slua_profileTabName);
#endif
}

void Fslua_profileModule::PluginButtonClicked()
{
#if (ENGINE_MINOR_VERSION<25) && (ENGINE_MAJOR_VERSION==4)
    FGlobalTabmanager::Get()->InvokeTab(slua_profileTabName);
#else
    FGlobalTabmanager::Get()->TryInvokeTab(slua_profileTabName);
#endif
}

bool Fslua_profileModule::Tick(float DeltaTime)
{
    if (!tabOpened)
    {
        return true;
    }
    //UE_LOG(LogTemp, Warning, TEXT("Fslua_profileModule::Tick"));
    while (!funcProfilerNodeQueue.IsEmpty())
    {
        MemoryFramePtr memoryFrame;
        memoryQueue.Dequeue(memoryFrame);
        TSharedPtr<FunctionProfileNode> funcProfilerNode;
        funcProfilerNodeQueue.Dequeue(funcProfilerNode);
        sluaProfilerInspector->Refresh(funcProfilerNode, memoryInfo, memoryFrame);
    }
    return true;
}

TSharedRef<class SDockTab> Fslua_profileModule::OnSpawnPluginTab(const FSpawnTabArgs& SpawnTabArgs)
{
    if (sluaProfilerInspector.IsValid())
    {
        sluaProfilerInspector->StartChartRolling();
        auto tab = sluaProfilerInspector->GetSDockTab();

        tab->SetOnTabClosed(SDockTab::FOnTabClosedCallback::CreateRaw(this, &Fslua_profileModule::OnTabClosed));
        SluaProfilerDataManager::InitProfileNode(funcProfilerRoot, ROOT_NAME, 0);
        sluaProfilerInspector->ProfileServer = MakeShareable(new NS_SLUA::FProfileServer());
        sluaProfilerInspector->ProfileServer->OnProfileMessageRecv().BindLambda([this](NS_SLUA::FProfileMessagePtr Message) {
            this->debug_hook_c(Message);
            });

        tabOpened = true;
        return tab;
    }
    else
    {
        return SNew(SDockTab).TabRole(ETabRole::NomadTab);
    }
}

//////////////////////////////////////////////////////////////////////////
#if WITH_EDITOR
Flua_profileCommands::Flua_profileCommands()
    : TCommands<Flua_profileCommands>(slua_profileTabName,
#if ENGINE_MAJOR_VERSION==5 && ENGINE_MINOR_VERSION>0
        NSLOCTEXT("Contexts", "slua_profile", "slua_profile Plugin"), NAME_None, FAppStyle::GetAppStyleSetName())
#else
        NSLOCTEXT("Contexts", "slua_profile", "slua_profile Plugin"), NAME_None, FEditorStyle::GetStyleSetName())
#endif
{
}

void Flua_profileCommands::RegisterCommands()
{
    UI_COMMAND(OpenPluginWindow, "slua Profile", "Open slua Profile tool", EUserInterfaceActionType::Button, FInputGesture());
}
#endif

void Fslua_profileModule::ClearCurProfiler()
{
    SluaProfilerDataManager::InitProfileNode(funcProfilerRoot, ROOT_NAME, 0);
    currentMemory = MakeShareable(new MemoryFrame());
    currentMemory->bMemoryTick = false;
    profilerStack.Empty();
}

void Fslua_profileModule::AddMenuExtension(FMenuBuilder& Builder)
{
#if WITH_EDITOR
    Builder.AddMenuEntry(Flua_profileCommands::Get().OpenPluginWindow);
#endif
}

void Fslua_profileModule::OnTabClosed(TSharedRef<SDockTab>)
{
    tabOpened = false;
}

void Fslua_profileModule::debug_hook_c(NS_SLUA::FProfileMessagePtr Message)
{
    //int event, double nanoseconds, int linedefined, const FString& name, const FString& short_src, TArray<NS_SLUA::LuaMemInfo> memoryInfoList
    int event = Message->Event;
    double nanoseconds = Message->Time;
    int linedefined = Message->Linedefined;
    const FString& name = Message->Name;
    const FString& short_src = Message->ShortSrc;
    if (event == NS_SLUA::ProfilerHookEvent::PHE_CALL)
    {
        if (linedefined == -1 && name.IsEmpty())
        {
            return;
        }
        FString functionName = short_src;
        functionName += ":";
        functionName += FString::FromInt(linedefined);
        functionName += " ";
        functionName += name;
        //UE_LOG(LogTemp, Log, TEXT("Profile Call %s"), *functionName);

        SluaProfilerDataManager::WatchBegin(functionName, nanoseconds, funcProfilerRoot, profilerStack);
    }
    else if (event == NS_SLUA::ProfilerHookEvent::PHE_RETURN)
    {
        if (linedefined == -1 && name.IsEmpty())
        {
            return;
        }
        
        FString functionName = short_src;
        functionName += ":";
        functionName += FString::FromInt(linedefined);
        functionName += " ";
        functionName += name;
        //UE_LOG(LogTemp, Log, TEXT("Profile Return %s"), *functionName);

        SluaProfilerDataManager::WatchEnd(functionName, nanoseconds, profilerStack);
    }
    else if (event == NS_SLUA::ProfilerHookEvent::PHE_TICK)
    {
        funcProfilerNodeQueue.Enqueue(funcProfilerRoot);
        memoryQueue.Enqueue(currentMemory);
        ClearCurProfiler();
    }
    else if (event == NS_SLUA::ProfilerHookEvent::PHE_MEMORY_TICK)
    {
        currentMemory->memoryInfoList = Message->memoryInfoList;
        currentMemory->bMemoryTick = true;
    }
    else if (event == NS_SLUA::ProfilerHookEvent::PHE_MEMORY_INCREACE)
    {
        currentMemory->memoryIncrease = Message->memoryIncrease;
    }
    else if (event == NS_SLUA::ProfilerHookEvent::PHE_ENTER_COROUTINE)
    {
        FString functionName = TEXT("");
        functionName += ":";
        functionName += FString::FromInt(linedefined);
        functionName += " ";
        functionName += name;

        //UE_LOG(LogTemp, Log, TEXT("Profile CoBegin %s"), *functionName);
        
        SluaProfilerDataManager::CoroutineBegin(functionName, nanoseconds, funcProfilerRoot, profilerStack);
    }
    else if (event == NS_SLUA::ProfilerHookEvent::PHE_EXIT_COROUTINE)
    {
        //UE_LOG(LogTemp, Log, TEXT("Profile CoEnd"));
        SluaProfilerDataManager::CoroutineEnd(nanoseconds, profilerStack);
    }
}

#undef LOCTEXT_NAMESPACE

IMPLEMENT_MODULE(Fslua_profileModule, slua_profile)

File: Content/Lua/slua_profile.lua
--[[
 Tencent is pleased to support the open source community by making sluaunreal available.

 Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
 Licensed under the BSD 3-Clause License (the "License");
 you may not use this file except in compliance with the License. You may obtain a copy of the License at

 https://opensource.org/licenses/BSD-3-Clause

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and limitations under the License.
]]

local ConsoleLogLevel = 2           --打印在控制台(print)的日志等级 0 : all/ 1: info/ 2: error.
local ConnectTimeoutSec = 0.005
local OpenAttachMode = true         --打开自动附加模式，随时可以连接profile
local AttachInterval = 1

local this = {}
local sock -- tcp socket
local connectHost
local connectPort
local currentRunState
local currentHookState
local stopConnectTime = 0
local profileTotalCost = 0

local ignoreHook = false

local RunState = {
    DISCONNECT = 0,
    CONNECTED = 1,
}

local HookState = {
    UNHOOK = 0,               --断开连接
    HOOK = 1,                 --全局无断点
};

local EventID = {
    ["call"] = 0,
    ["return"] = 1,
}

local function getTime()
    return slua.getNanoseconds()
end

function this.start(host, port)
    host = tostring(host or "127.0.0.1")
    port = tonumber(port) or 8081
    this.printToConsole("Profile start. connect host:" .. host .. " port:".. tostring(port), 1)
    if sock ~= nil then
        this.printToConsole("[Warning] Profiler已经启动，请不要再次调用start()" , 1)
        return
    end

    --尝试初次连接
    this.changeRunState(RunState.DISCONNECT)
    this.reGetSock()
    connectHost = host
    connectPort = port
    local sockSuccess = sock and sock:connect(connectHost, connectPort)
    if sockSuccess ~= nil then
        this.printToConsole("first connect success!")
        this.connectSuccess()
    else
        this.printToConsole("first connect failed!")
        this.changeHookState(HookState.UNHOOK)
    end
end

-- 把日志打印在控制台
-- @str: 日志内容
-- @printLevel: all(0)/info(1)/error(2)
function this.printToConsole(str, printLevel)
    printLevel = printLevel or 0
    if ConsoleLogLevel > printLevel then
        return
    end
    print("[Slua Profile] ".. tostring(str))
end

function this.changeRunState(state)
    currentRunState = state
end

function this.changeHookState(state)
    currentHookState = state

    if currentHookState == HookState.UNHOOK then
        profileTotalCost = 0
        if OpenAttachMode == true then
            debug.sethook(this.debug_hook, "r", 1000000)
        else
            debug.sethook()
        end
    elseif currentHookState == HookState.HOOK then
        debug.sethook(this.debug_hook, "rc")
    end
end

function this.reGetSock()
    if sock ~= nil then
        pcall(function() sock:close() end)
    end

    if pcall(function() sock =  require("socket.core").tcp() end) then
        this.printToConsole("reGetSock success")
        sock:settimeout(ConnectTimeoutSec)
    else
        this.printToConsole("[Error] reGetSock fail", 2)
    end
end

local function startsWith(source, start)
    return string.sub(source,1,string.len(start))== start
end

function this.debug_hook(event)
    local start = getTime()
    if ignoreHook then
        return
    end

    if this.reConnect() then
        return
    end

    local eventID = EventID[event]
    if eventID == nil then
        return
    end

    local info = debug.getinfo(2, "nSl")
    if info.short_src and startsWith(info.short_src, "slua_profile.lua") then
        return
    end

    local block = slua.makeProfilePackage(eventID, start - profileTotalCost, info.linedefined, info.name or "", info.short_src or "")
    this.sendMsg(block)

    profileTotalCost = profileTotalCost + (getTime() - start)
end

function this.sendMsg(msg)
    local succ,err;
    if pcall(function() succ,err = sock:send(msg) end) then
        if succ == nil then
            if err == "closed" then
                this.disconnect();
            end
        end
    end
end

-- 定时(以函数return为时机) 进行attach连接
function this.reConnect()
    if currentHookState == HookState.UNHOOK then
        if os.time() - stopConnectTime < AttachInterval then
            this.printToConsole("Reconnect time less than 1s")
            this.printToConsole("os.time:".. os.time() .. " | stopConnectTime:" ..stopConnectTime)
            return true
        end

        if not sock then
            this.reGetSock()
        end

        local sockSuccess, status = sock:connect(connectHost, connectPort)
        if sockSuccess == 1 or status == "already connected" then
            this.printToConsole("reconnect success")
            this.connectSuccess()
        else
            this.printToConsole("reconnect failed . retCode:" .. tostring(sockSuccess) .. "  status:" .. status)
            stopConnectTime = os.time()
        end

        return true
    end

    return false
end

function this.connectSuccess()
    this.changeRunState(RunState.CONNECTED)
    this.printToConsole("connectSuccess", 1)
    this.changeHookState(HookState.HOOK)
end

function this.disconnect()
    this.printToConsole("Profiler disconnect", 1)

    this.changeHookState( HookState.UNHOOK )
    stopConnectTime = os.time()
    this.changeRunState(RunState.DISCONNECT)

    if sock ~= nil then
        sock:close()
    end
    this.reGetSock()
end

local function profileTick(deltaTime)
    ignoreHook = true
    if currentRunState == RunState.CONNECTED then
        local block = slua.makeProfilePackage(-1, getTime(), -1, "", "")
        this.sendMsg(block)
    end
    ignoreHook = false
end

slua.setTickFunction(profileTick)

return this